Loader: Intro, `fabric.mod.json`, Dependency Overrides (#557)
FabricMC/fabric-docs: develop/loader/index.md

---
title: Fabric Loader
description: The lightweight mod loader that powers the Fabric project.
authors:
  - cassiancc
  - falseresync
  - jamieswhiteshirt
  - Llamalad7
authors-nogithub:
  - liach

>   - solidblock

Reviewer: I think this is `SolidBlock-cn`?

Author: Could be - if you still have that wiki author sheet that'd be handy.

---

> Fabric Loader is Fabric's lightweight mod loader. It provides the necessary tools to make Minecraft modifiable without depending on a specific version of the game. Game specific (and game version specific) hooks belong in Fabric API. It is possible to adapt Fabric Loader for many Java applications (for instance games like Slay the Spire and Starmade).

Reviewer: ```suggestion
Fabric Loader is Fabric's lightweight mod loader. It provides the necessary tools to make any version of Minecraft. On the other hand, game-specific and version-specific hooks belong in Fabric API.

::: info

It is possible to adapt Fabric Loader for many Java applications (for instance games like Slay the Spire and Starmade).

:::
```

- Do we have a link to a page about Fabric API?
- Is Loader really compatible with any version of MC?

Author: 1. I don't believe so.
2. Yes, though versions below 1.14 would require third party intermediary mappings.

> Fabric Loader has services to allow mods to have some code executed during initialization, to transform classes, declare and provide mod dependencies, all in a number of different environments.

Reviewer: ```suggestion
Fabric Loader has services that allow mods to execute code during initialization, transform classes, declare and provide mod dependencies, all that in various environments.
```

> The Javadoc for the latest vesrsion of Fabric Loader's can be found on the [Develop](https://fabricmc.net/develop/) site.

Reviewer: ```suggestion
The Javadocs for the latest version of Fabric Loader can be found on the [Develop](https://fabricmc.net/develop/) site.
```
I think it's javadocs with an s, no?

> The current instance of Fabric Loader can be retrieved by using `FabricLoader.getInstance()`. As an example, `FabricLoader.getInstance().isModLoaded` can be used to check for the presence of another running mod.

Reviewer: ```suggestion
The current instance of Fabric Loader can be retrieved by using `FabricLoader.getInstance()`. As an example, `FabricLoader.getInstance().isModLoaded` can be used to check whether another mod is running.
```

## Mods {#mods}

> A mod is a jar with a [`fabric.mod.json`](./fabric-mod-json) mod metadata file in its root declaring how it should be loaded. It primarily declares a mod ID and version as well as [entrypoints](https://wiki.fabricmc.net/documentation:entrypoint) and mixin configurations. The mod ID identifies the mod so that any mod with the same ID is considered to be the same mod. Only one version of a mod may be loaded at a time. A mod may declare other mods that it depends on or conflicts with. Fabric Loader will attempt to satisfy dependencies and load the appropriate versions of mods, or fail to launch otherwise.

Reviewer: ```suggestion
A mod is a jar with a [`fabric.mod.json`](./fabric-mod-json) metadata file in its root declaring how it should be loaded. That file contains a mod ID and version number, as well as [entrypoints](https://wiki.fabricmc.net/documentation:entrypoint) and mixin configurations.

The mod ID identifies the mod, and any two mods with the same ID are considered to be the same mod. Only one version of a mod may be loaded at a time.

A mod may also declare which other mods it depends on or conflicts with. Fabric Loader will attempt to satisfy dependencies and load the appropriate versions of mods; if it cannot it will fail to launch the game.
```

we have a section in the docs about entrypoints I think. I'd prefer if this linked to the Docs, and the wiki link were in the other page.

Also, I haven't read the other page yet in this review, but I have a feeling this information will be duplicated; regardless, maybe it should be moved there.

> Fabric Loader makes all mods equally capable of modifying the game. As an example, anything Fabric API does can be done by any other mod.
> 

Reviewer: ```suggestion
```

(why) do you think this needs to be specified?

Author: I believe this was included as a comparison point to loaders like Forge that do not grant mod authors the same capabilities Fabric does.

Mods are loaded both from the classpath and from the `mods` directory. This directory can be changed with the `fabric.modsFolder` system property.

## Nested JARs {#nested-jars}

> Nested JARs allow a mod to provide its own dependencies, so Fabric Loader can pick the best version matching the dependencies instead of requiring separate installation of dependencies. They also allow clean packaging of submodules, so each module can be used separately. Non-mod libraries can be repackaged as mods for nested JAR usage. A mod may bundle a number of other mods within its JAR. A nested JAR must itself also be a mod, which again can have nested JARs. Fabric Loader will load nested JARs while attempting to satisfy dependency constraints.

Reviewer: ```suggestion
Nested JARs allow a mod to provide its own dependencies, so Fabric Loader can find the best version while attempting to satisfy dependencies, instead of requiring them to be installed separately.

They essentially are like any regular mod, with their own metadata file, but are contained within the parent JAR. Note that nested mods may themselves embed other children in the same way.

They also allow clean packaging of submodules, so each module can be used separately. For example, non-mod libraries could be repackaged as mods and shipped as nested JARs.
```

are these the same thing as jar-in-jar? if so I'd note that. I also vaguely remember that those were mentioned somewhere in the docs? don't quote me on that please

> so each module can be used separately

not so sure what point this is making. maybe it means "developed"?

Author: Yes, this is the same thing. It might be detailed in the Loom docs, I'll take a look through those again.

That sentence really just needs rewording, I agree it's not very clear what the benefits are.

> Nested JARs are extracted to the disk when the game is run. Nested JARs must be declared by their paths relative to the containing JAR's root.

Reviewer: ```suggestion
Nested JARs are extracted to the disk when the game is run. Paths to nested JARs must be declared relative to their containing JAR's root.
```

> Using Fabric Loom's `include` option will automatically handle nesting the jar.

Reviewer: ```suggestion
Using Fabric Loom's `include` option will automatically handle nesting the JAR.
```

`https://maven.fabricmc.net/docs/fabric-loader-[loader version]`
For example, Fabric Loader 0.18.6's documentation is at https://maven.fabricmc.net/docs/fabric-loader-0.18.6/

The current instance of Fabric Loader can be retrieved by using `FabricLoader.getInstance()`. As an example, `FabricLoader.getInstance().isModLoaded` can be used to check for the presence of another running mod.

## Mods {#mods}

A mod is a jar with a [`fabric.mod.json`](./fabric-mod-json) mod metadata file in its root declaring how it should be loaded. It primarily declares a mod ID and version as well as [entrypoints](https://wiki.fabricmc.net/documentation:entrypoint) and mixin configurations. The mod ID identifies the mod so that any mod with the same ID is considered to be the same mod. Only one version of a mod may be loaded at a time. A mod may declare other mods that it depends on or conflicts with. Fabric Loader will attempt to satisfy dependencies and load the appropriate versions of mods, or fail to launch otherwise.

Fabric Loader makes all mods equally capable of modifying the game. As an example, anything Fabric API does can be done by any other mod.

> Mods are loaded both from the classpath and from the mods directory. They are expected to match the mappings in the current environment, meaning Fabric Loader will not remap any mods.

Reviewer 2: > They are expected to match the mappings in the current environment, meaning Fabric Loader will not remap any mods.

Not correct, if loader has the correct mappings in dev it will remap mods. Eg. in dev you can put intermediary mods into /mods and loader will remap them to named.

## Nested JARs {#nested-jars}

Nested JARs allow a mod to provide its own dependencies, so Fabric Loader can pick the best version matching the dependencies instead of requiring separate installation of dependencies. They also allow clean packaging of submodules, so each module can be used separately. Non-mod libraries can be repackaged as mods for nested JAR usage. A mod may bundle a number of other mods within its JAR. A nested JAR must itself also be a mod, which again can have nested JARs. Fabric Loader will load nested JARs while attempting to satisfy dependency constraints.

> Nested JARs are not extracted, they are instead loaded in in-memory file system using jimfs. Nested JARs must be declared by their paths relative to the containing JAR's root.

Reviewer 2: No longer correct as of many years ago, loader extracts the jar to the disk.

Mixin was not originally made for Fabric, so Fabric Loader uses a modified version of Mixin. However, the documentation of the upstream version is still mostly valid. Fabric's modifications include allowing all default injection points inside constructors, optimizing out unused callback infos, providing fixes for backwards compatibility, fixing static shadows, allowing injectors in interfaces, and more.

<!-- Referenced comments from LlamaLad7 on Fabric Mixin. https://discord.com/channels/507304429255393322/566418023372816394/1002211121903706162 -->

## Mappings {#mappings}

:::info

Mappings are only relevant when using Fabric Loader on obfuscated games, including pre-26.1 versions of Minecraft.

:::

Fabric Loader provides the `MappingResolver` API to determine names of classes, fields and methods with respect to the different environments that mods may be loaded in. This can be used to support reflection in any environment provided Fabric Loader has access to mappings to resolve the name.

```java
FabricLoader.getInstance().getMappingResolver().mapClassName("intermediary", "net.minecraft.class_5421") // Resolves to `RecipeBookType` on named versions
```

> When launched in a non-development environment on an obfuscated game, Fabric Loader will [remap](/porting/mappings/#mappings) the Minecraft jar and Realms client jar to intermediary names. Mods designed for obfuscated games are expected to be mapped to intermediary, which will be compatible with this environment. The remapped jars are cached and saved in `${gameDir}/.fabric/remappedJars/${minecraftVersion}` for re-use across launches.

Reviewer 2: I dont think mentioning the realms jar is that useful, it was merged into the client jar many years ago. It still does it for versions that requires it, just seems odd to mention it here.
